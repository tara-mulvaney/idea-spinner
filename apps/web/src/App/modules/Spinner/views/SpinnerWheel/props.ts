export interface SpinnerWheelProps {
  description?: string;
  isLocked: boolean;
  isSpinning: boolean;
  name: string;
  value?: string;
}

export default {
  description:
    "Clowns are a curious breed who sustain themselves " +
    "on the laugher of children.",
  isLocked: false,
  isSpinning: false,
  name: "Demo Wheel",
  value: "Clowns",
} as SpinnerWheelProps;
