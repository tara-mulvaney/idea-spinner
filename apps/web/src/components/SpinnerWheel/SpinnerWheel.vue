<script lang="ts" setup>
  import { ref, watch } from "vue";

  // see: https://github.com/vuejs/core/issues/4294
  interface SpinnerWheelProps {
    description?: string;
    isLocked: boolean;
    isSpinning: boolean;
    name: string;
    value?: string;
  }

  const props = defineProps<SpinnerWheelProps>();

  const input = ref<HTMLInputElement | null>(null);

  const isAnimating = ref(false);
  const hasInnerFocus = ref(false);

  const displayEmptyValue = "--";

  watch(
    () => props.value,
    () => {
      isAnimating.value = true;
    }
  );

  const emit = defineEmits(["edit", "lock", "unlock"]);

  function toggleLock() {
    if (props.isSpinning) {
      return;
    }

    emit(props.isLocked ? "unlock" : "lock", props.name);
  }

  function updateValue({ target }: Event) {
    const { value } = target as HTMLInputElement;

    emit("edit", { value });
  }

  function _resetAnimation() {
    isAnimating.value = false;
  }

  function _delegateKeystroke(event: KeyboardEvent) {
    switch (event.key) {
      case "Enter":
        event.preventDefault();
        return toggleLock();
      case "Tab":
        return;
      default:
        input.value?.focus();
    }
  }
</script>

<template>
  <li
    :class="{
      SpinnerWheel: true,
      'SpinnerWheel--stopped': isLocked || !isSpinning,
      'SpinnerWheel--tick': !isLocked && isAnimating && isSpinning,
      'SpinnerWheel--innerFocus': hasInnerFocus,
    }"
    :tabindex="value === undefined ? -1 : 1"
    @animationend="_resetAnimation()"
    @keydown="_delegateKeystroke($event)"
  >
    <header class="SpinnerWheel__header">
      <label
        class="SpinnerWheel__name"
        :for="name"
      >
        {{ name }}
      </label>
      <i
        :class="{
          SpinnerWheel__lock: true,
          'SpinnerWheel__lock--locked': isLocked,
        }"
        role="button"
        tabindex="-1"
        @click.prevent="toggleLock()"
        @keydown.enter="toggleLock()"
      >
        {{ isLocked ? "🔒" : "🔓" }}
      </i>
    </header>
    <button
      :class="{
        SpinnerWheel__value: true,
        'SpinnerWheel__value--description':
          Boolean(description) && (isLocked || !isSpinning),
      }"
      tabindex="-1"
      :title="description"
      @click.stop.prevent="input?.focus()"
      @keydown="input?.focus()"
    >
      <!--
        ISSUE #32: somehow get contenteditable to work

        For the life of me I could not get contenteditable 
        to work in Vue3 - so, my clever workaround was to bind the 
        size attribute to the current input value's length.

        Still, contenteditable would allow for text wrapping here,
        which would be nice (textarea doesn't work)
      -->
      <input
        :id="name"
        ref="input"
        class="SpinnerWheel__valueInput"
        :disabled="value === undefined || isSpinning"
        :size="value?.length ?? displayEmptyValue.length"
        tabindex="-1"
        :value="value ?? displayEmptyValue"
        @blur="hasInnerFocus = false"
        @focus="hasInnerFocus = true"
        @input="updateValue($event)"
      />
    </button>
  </li>
</template>

<style scoped>
  * {
    all: initial;
  }

  .SpinnerWheel {
    animation-duration: var(--animation-timing);
    background: var(--color-primary);
    border-radius: var(--gutter-narrow);
    flex-direction: column;
    height: 100%;
    overflow: hidden;
    position: relative;
    transition: background var(--animation-timing) ease-out;
  }

  .SpinnerWheel,
  .SpinnerWheel__header,
  .SpinnerWheel__value {
    align-items: center;
    display: flex;
    justify-content: center;
  }

  .SpinnerWheel:focus,
  .SpinnerWheel--innerFocus {
    outline: 3px solid var(--color-primary-variant);
    outline-offset: 2px;
  }

  .SpinnerWheel__header {
    background: var(--color-primary-variant);
    border-bottom: var(--hairline) solid var(--color-primary);
    height: var(--gutter-wide);
    line-height: var(--gutter-wide);
    width: 100%;
  }

  .SpinnerWheel__header * {
    user-select: none;
  }

  .SpinnerWheel__name {
    color: var(--color-primary-variant-foreground);
    font-family: var(--font-primary);
    font-size: var(--gutter-standard);
    font-weight: bold;
    text-transform: uppercase;
  }

  .SpinnerWheel__lock {
    cursor: pointer;
    float: right;
    line-height: var(--gutter-wide);
    opacity: 0.5;
    position: absolute;
    right: var(--gutter-standard);
  }

  .SpinnerWheel__lock.SpinnerWheel__lock--locked {
    opacity: 1;
  }

  .SpinnerWheel__value {
    box-sizing: border-box;
    flex-grow: 1;
    padding: var(--gutter-narrow);
    width: 100%;
  }

  .SpinnerWheel__valueInput {
    color: var(--color-primary-foreground);
    font-family: var(--font-secondary);
    font-size: var(--gutter-large);
    max-width: 100%;
    text-align: center;
    text-overflow: ellipsis;
    transition: color var(--animation-timing) ease-out;
  }

  .SpinnerWheel__value--description {
    cursor: help;
  }

  .SpinnerWheel__value--description::before {
    content: "ℹ️";
    padding: 0 var(--gutter-narrow);
  }

  .SpinnerWheel--tick {
    animation-name: tick;
  }

  @keyframes tick {
    from {
      background-color: var(--color-primary-variant);
      border-color: var(--color-primary-variant-foreground);
      color: var(--color-primary-variant-foreground);
    }

    to {
      background-color: var(--color-primary);
      border-color: var(--color-primary-foreground);
      color: var(--color-primary-foreground);
    }
  }

  .SpinnerWheel--stopped {
    background: var(--color-primary-variant);
  }

  .SpinnerWheel--stopped .SpinnerWheel__valueInput {
    color: var(--color-primary-variant-foreground);
  }

  .SpinnerWheel--stopped > .SpinnerWheel__name {
    border-color: var(--color-primary-variant-foreground);
  }
</style>
