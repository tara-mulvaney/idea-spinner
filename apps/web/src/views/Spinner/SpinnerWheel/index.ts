export type { SpinnerWheelProps } from "./props";
export { default as SpinnerWheel } from "./SpinnerWheel.vue";
