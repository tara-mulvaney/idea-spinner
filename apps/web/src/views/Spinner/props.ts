import type { SpinnerWheelProps } from "./SpinnerWheel";

export interface SpinnerProps {
  hasSpun: boolean;
  isLocked: boolean;
  isSpinning: boolean;
  maxColumns?: number;
  wheels: SpinnerWheelProps[];
}

export default {
  hasSpun: false,
  isLocked: false,
  isSpinning: false,
  maxColumns: 4,
  wheels: [
    {
      isLocked: false,
      isSpinning: true,
      name: "color",
      value: "red",
    },
    {
      isLocked: false,
      isSpinning: true,
      name: "shape",
      value: "round",
    },
    {
      isLocked: false,
      isSpinning: false,
      name: "feeling",
      value: "sad",
    },
  ],
} as SpinnerProps;
