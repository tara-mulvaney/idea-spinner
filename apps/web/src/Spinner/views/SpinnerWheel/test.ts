import props from "./props";
import SpinnerWheel from "./SpinnerWheel.vue";
import { expect, test } from "@jest/globals";
import { mount, shallowMount } from "@vue/test-utils";

test.concurrent("spinner wheel - snapshot", async () => {
  const { element } = shallowMount(SpinnerWheel, { props });

  expect(element).toMatchSnapshot();
});

const TOP = ".SpinnerWheel";
const LOCK = ".SpinnerWheel__lock";
const VALUE_INPUT = ".SpinnerWheel__valueInput";

test.concurrent(`spinner wheel - ${LOCK}`, async () => {
  const wrapper = mount(SpinnerWheel, { props });

  await wrapper.find(LOCK).trigger("click");

  expect(wrapper.emitted()).toHaveProperty("lock");
  expect(wrapper.emitted()).not.toHaveProperty("unlock");
});

test.concurrent(`spinner wheel - ${TOP} Enter`, async () => {
  const wrapper = mount(SpinnerWheel, { props });

  await wrapper.find(TOP).trigger("keydown", { key: "Enter" });

  expect(wrapper.emitted()).toHaveProperty("lock");
  expect(wrapper.emitted()).not.toHaveProperty("unlock");
});

test.concurrent(
  `spinner wheel - ${LOCK} shouldn't toggle while spinning`,
  async () => {
    const wrapper = mount(SpinnerWheel, {
      props: { ...props, isSpinning: true },
    });

    await wrapper.find(LOCK).trigger("click");

    expect(wrapper.emitted()).not.toHaveProperty("lock");
  }
);

test.concurrent(
  `spinner wheel - tabbing from ${TOP} shouldn't change state`,
  async () => {
    const wrapper = mount(SpinnerWheel, { props });

    await wrapper.find(TOP).trigger("keydown", { key: "Tab" });

    expect(wrapper.emitted()).not.toHaveProperty("lock");
    expect(wrapper.emitted()).not.toHaveProperty("edit");
  }
);

test.concurrent(`spinner wheel - change ${VALUE_INPUT}`, async () => {
  const wrapper = mount(SpinnerWheel, { props });

  await wrapper.find(VALUE_INPUT).trigger("input");

  expect(wrapper.emitted()).toHaveProperty("edit");
});

test.concurrent(
  `spinner wheel - change ${VALUE_INPUT} from top-level ${TOP}`,
  async () => {
    const wrapper = mount(SpinnerWheel, { props });

    await wrapper.find(TOP).trigger("keydown", { keyCode: 65 });

    expect(wrapper.emitted()).toHaveProperty("edit");
  }
);
