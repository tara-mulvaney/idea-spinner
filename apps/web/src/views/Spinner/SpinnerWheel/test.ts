import props from "./props";
import SpinnerWheel from "./SpinnerWheel.vue";
import { expect, test } from "@jest/globals";
import { mount, shallowMount } from "@vue/test-utils";

test.concurrent("spinner wheel - snapshot", async () => {
  const { element } = shallowMount(SpinnerWheel, { props });

  expect(element).toMatchSnapshot();
});

const TOP = ".SpinnerWheel";
const LOCK = ".SpinnerWheel__lock";
const VALUE_INPUT = ".SpinnerWheel__valueInput";

test.concurrent(`spinner wheel - ${LOCK}`, async () => {
  const wrapper = mount(SpinnerWheel, { props });

  await wrapper.find(LOCK).trigger("click");

  expect(wrapper.emitted()).toHaveProperty("lock");
  expect(wrapper.emitted()).not.toHaveProperty("unlock");
});

test.concurrent(`spinner wheel - ${LOCK}, unlock`, async () => {
  const wrapper = mount(SpinnerWheel, { props: { ...props, isLocked: true } });

  await wrapper.find(LOCK).trigger("click");

  expect(wrapper.emitted()).not.toHaveProperty("lock");
  expect(wrapper.emitted()).toHaveProperty("unlock");
});

test.concurrent(`spinner wheel - ${TOP} Enter`, async () => {
  const wrapper = mount(SpinnerWheel, { props });

  await wrapper.find(TOP).trigger("keydown", { key: "Enter" });

  expect(wrapper.emitted()).toHaveProperty("lock");
  expect(wrapper.emitted()).not.toHaveProperty("unlock");
});

test.concurrent(
  `spinner wheel - ${LOCK} shouldn't toggle while spinning`,
  async () => {
    const wrapper = mount(SpinnerWheel, {
      props: { ...props, isSpinning: true },
    });

    await wrapper.find(LOCK).trigger("click");

    expect(wrapper.emitted()).not.toHaveProperty("lock");
    expect(wrapper.emitted()).not.toHaveProperty("unlock");
  }
);

test.concurrent(
  `spinner wheel - tabbing/editing from ${TOP} shouldn't change state directly`,
  async () => {
    const wrapper = mount(SpinnerWheel, { props });

    await wrapper.find(TOP).trigger("keydown", { key: "Tab" });
    await wrapper.find(TOP).trigger("keydown", { keyCode: 65 });

    expect(wrapper.emitted()).not.toHaveProperty("lock");
    expect(wrapper.emitted()).not.toHaveProperty("edit");
  }
);

test.concurrent(`spinner wheel - change ${VALUE_INPUT}`, async () => {
  const wrapper = mount(SpinnerWheel, { props });

  await wrapper.find(VALUE_INPUT).trigger("input");

  expect(wrapper.emitted()).toHaveProperty("edit");
});
