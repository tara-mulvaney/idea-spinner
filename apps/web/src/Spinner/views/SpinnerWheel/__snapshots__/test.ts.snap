// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `
<li
  class="SpinnerWheel SpinnerWheel--stopped"
  tabindex="1"
>
  <header
    class="SpinnerWheel__header"
  >
    <label
      class="SpinnerWheel__name"
      for="Demo Wheel"
    >
      Demo Wheel
    </label>
    <i
      class="SpinnerWheel__lock"
      role="button"
      tabindex="-1"
    >
      ðŸ”“
    </i>
  </header>
  <button
    class="SpinnerWheel__value SpinnerWheel__value--description"
    tabindex="-1"
    title="Clowns are a curious breed who sustain themselves on the laugher of children."
  >
    <!--
        ISSUE #32: somehow get contenteditable to work

        For the life of me I could not get contenteditable 
        to work in Vue3 - so, my clever workaround was to bind the 
        size attribute to the current input value's length.
      -->
    <input
      class="SpinnerWheel__valueInput"
      id="Demo Wheel"
      size="6"
      tabindex="-1"
    />
  </button>
</li>
`;
